Extensions/TweenBehavior/tweenruntimebehavior.js

```diff
@@ -0,0 +1,640 @@
+};
+
+gdjs.TweenRuntimeBehavior.prototype._getTween = function(identifier) {
+  return this._tweens[identifier];
+};
+
+gdjs.TweenRuntimeBehavior.prototype._tweenExists = function(identifier){ 
+  return !!this._tweens[identifier]; 
+}
+
+gdjs.TweenRuntimeBehavior.prototype._tweenIsPlaying = function(identifier) {
+  return this._tweens[identifier].instance.isPlaying();
+};
+
+gdjs.TweenRuntimeBehavior.prototype._pauseTween = function(identifier) {
+  return this._tweens[identifier].instance.pause();
+};
+
+gdjs.TweenRuntimeBehavior.prototype._resumeTween = function(identifier) {
+  return this._tweens[identifier].instance.resume();
+};
+
+gdjs.TweenRuntimeBehavior.prototype._stopTween = function(identifier, JumpToDest) {
+  return this._tweens[identifier].instance.stop(JumpToDest);
+};
+
+gdjs.TweenRuntimeBehavior.prototype._setTweenFinished = function(identifier, hasFinished) {
+  this._tweens[identifier].hasFinished = hasFinished;
+};
+
+gdjs.TweenRuntimeBehavior.prototype._tweenHasFinished = function(identifier) {
+  return this._tweens[identifier].hasFinished;
+};
+
+gdjs.TweenRuntimeBehavior.prototype._removeObjectFromScene = function(identifier) {
+  if (this._tweenExists(identifier)){
+    this._removeTween(identifier)
+  }
+  return this.owner.deleteFromScene(this.runtimeScene);
+};
+
+gdjs.TweenRuntimeBehavior.prototype._removeTween = function(identifier) {
+  this._tweens[identifier].instance.dispose();
+  delete this._tweens[identifier];
+};
+
+/**
+ * Add an object variable tween.
+ * @param {string} identifier Unique id to idenfify the tween
+ * @param {objectvar} variable The object variable to store the tweened value
+ * @param {number} fromValue Start value
+ * @param {number} toVaue End value
+ * @param {string} easingValue Type of easing
+ * @param {number} durationValue Duration in milliseconds
+ * @param {boolean} destroyObjectWhenFinished Destroy this object when the tween ends
+ */
+gdjs.TweenRuntimeBehavior.prototype.addObjectVariableTween = function (
+  identifier,
+  variable,
+  fromValue,
+  toValue,
+  easingValue,
+  durationValue,
+  destroyObjectWhenFinished
+) {
+
+  if (this._tweenExists(identifier)) {
+    this.removeTween(identifier);
+  }
+
+  var newTweenable = new shifty.Tweenable();
+
+  newTweenable.setConfig({
+    from: { value: fromValue, progress: 0.0 },
+    to: { value: toValue, progress: 1.0 },
+    duration: durationValue,
+    easing: easingValue,
+    step: state => {
+      //var instanceVariable = this.owner.getVariables().get(variable);
+      this._getTween(identifier).progress = state.progress;
+      this._getTween(identifier).currentValue = state.value;
+      //this.owner.setVariableNumber(instanceVariable, state.value) 
+      this.owner.setVariableNumber(variable, state.value);
+      console.log(state);
+    }
+  });
+
+  this._addObjectVariableTween(identifier, newTweenable, variable, fromValue);
+
+  if (destroyObjectWhenFinished) {
```

Comment:
I thought that I had already mentioned this, this code is repeated in every function:

```
  if (destroyObjectWhenFinished) {
    this._tweens[identifier].instance.tween().then(() => {
       this._removeObjectFromScene(identifier);
     });
  } else {
    this._tweens[identifier].instance.tween().then(() => {
      this._setTweenFinished(identifier, true);
    });
  }
```

It looks like an ideal candidate for being moved to a tool method that you then just call. Will shorten methods and make everything easier to read, and maintain :)

Reply:
Yes, you did :wink:
I have already turned it into a function in my working copy.